packages/contracts-bedrock/src/L1/DelayedVetoable.sol

```diff
@@ -0,0 +1,199 @@
+    function queuedAt(bytes32 callHash) external readOrHandle returns (uint256 queuedAt_) {
+        queuedAt_ = _queuedAt[callHash];
+    }
+
+    /// @notice Used for all calls that pass data to the contract.
+    fallback() external {
+        _handleCall();
+    }
+
+    /// @notice Receives all calls other than those made by the vetoer.
+    ///         This enables transparent initiation and forwarding of calls to the target and avoids
+    ///         the need for additional layers of abi encoding.
+    function _handleCall() internal {
+        // The initiator and vetoer activate the delay by passing in null data.
+        if (msg.data.length == 0) {
+            if (msg.sender != INITIATOR && msg.sender != VETOER) {
+                // todo(maurelian): make this error have an expected array.
+                revert Unauthorized(INITIATOR, msg.sender);
+            }
+            _activateDelay();
+            return;
+        }
+
+        bytes32 callHash = keccak256(msg.data);
+
+        // Case 1: The initiator is calling the contract to initiate a call.
+        if (msg.sender == INITIATOR && _queuedAt[callHash] == 0) {
+            if (_delay == 0) {
+                // This forward function will halt the call frame on completion.
+                _forwardAndHalt(callHash);
+            }
+            _queuedAt[callHash] = block.timestamp;
+            emit Initiated(callHash, msg.data);
+            return;
+        }
+
+        // Case 2: The vetoer is calling the contract to veto a call.
+        // Note: The vetoer retains the ability to veto even after the delay has passed. This makes censoring the vetoer
+        //       more costly, as there is no time limit after which their transaction can be included.
+        if (msg.sender == VETOER && _queuedAt[callHash] != 0) {
+            delete _queuedAt[callHash];
+            emit Vetoed(callHash, msg.data);
+            return;
+        }
+
+        // Case 3: The call is from an unpermissioned actor. We'll forward the call if the delay has
+        // passed.
+        if (_queuedAt[callHash] == 0) {
+            // The call has not been initiated, so we'll treat this is an unauthorized initiation attempt.
+            revert Unauthorized(INITIATOR, msg.sender);
+        }
+
+        if (_queuedAt[callHash] + _delay < block.timestamp) {
+            // Not enough time has passed, so we'll revert.
+            revert ForwardingEarly();
+        }
+
+        // Delete the call to prevent replays
+        delete _queuedAt[callHash];
+        _forwardAndHalt(callHash);
+    }
+
+    /// @notice Forwards the call to the target and halts the call frame.
+    function _forwardAndHalt(bytes32 callHash) internal {
+        // Forward the call
+        emit Forwarded(callHash, msg.data);
+        (bool success,) = TARGET.call(msg.data);
+        assembly {
+            // Success == 0 means a revert. We'll revert too and pass the data up.
```

Comment:
This passes the data up? The call will copy the memory back but not at the zero'th offset?

Reply:
Oh, good catch.  
I fixed this and made sure the tests assert on the returndata:
https://github.com/ethereum-optimism/optimism/blob/e2c0949d0496edb7937e6ca5446c7713884d0a90/packages/contracts-bedrock/test/DelayedVetoable.t.sol#L101
https://github.com/ethereum-optimism/optimism/blob/e2c0949d0496edb7937e6ca5446c7713884d0a90/packages/contracts-bedrock/test/DelayedVetoable.t.sol#L177-L179